File: array/java/TrappingRainWater.java
public class TrappingRainWater {
    int trap(int[] height){
        if (height.length == 0){
            return 0;
        }
        int res = 0;
        int left[] = new int[height.length];
        int right[] = new int[height.length];
        left[0] = height[0];
        right[height.length-1] = height[height.length-1];
        for (int i = 1; i<height.length; i++){
            left[i] = Math.max(left[i-1], height[i]);
        } 
        for (int i = height.length-2; i>-1; i--){
            right[i] = Math.max(right[i+1], height[i]);
        } 
        for (int i = 0; i<height.length; i++){
            res = Math.min(left[i],right[i]-height[i]);
        } 
        return res;
    }

    public static void main(String[] args){
        TrappingRainWater trw = new TrappingRainWater();
        int height[] = new int[]{1,2,3,4,5,6,5,4,3,2,1};
        int res = trw.trap(height);
        System.out.println("Traped water: "+res);
    }
}

File: array/python/TrappingRainWater.py
def trap(height):
    if not height:  return 0
    n, res = len(height), 0
    left, right = [0] * n, [0] * n
    left[0] = height[0]
    right[-1] = height[-1]
    for i in range(1, n):
        left[i] = max(left[i - 1], height[i])
    for i in range(n - 2, -1, -1):
        right[i] = max(right[i + 1], height[i])
    for i in range(n):
        res += min(left[i], right[i]) - height[i]
    return res

if __name__=='__main__':
    test_cases = [[0,1,0,2,1,0,1,3,2,1,2,1],
                  [4,2,0,3,2,5]]
    for tc in test_cases:
        print(trap(tc))